### ISSUE TITLE
WCF and Load Balanced Environment

### ISSUE THREAD

**Issue by mieliespoor** [App-vNext/Polly#182]
We implemented Polly in a WCF Service and it works well in a normal single server environment. It seems to be giving issues in our production environment which is load balanced. I don't have a load balanced test environment I can test this with at the moment, but will get it sorted in a few days.

Are there any extra measures I need to implement to have this working in a load balanced environment?

Our production support team says that the app (in IIS) 'hangs' and then causes timeout issues.

**Comment by reisenberger** [comment]
Hi @mieliespoor
Not aware of any extra steps needed in a load-balanced environment, and not had similar reports.  (But _very_ happy to work with you further on this if we can identify particular causes.)

+ Have you been able to isolate Polly as the cause?  (ie eliminate that it is not related to anything else which may have changed in the production environment or codebase?).  
  - One approach to focus on this could be introduce a config feature-toggle switch to either pass the executions through Polly or not (with codebase otherwise identical).  Then we could see if the problem specifically manifests itself (or not) when the executions are passed through Polly or not.  (Just to isolate definitively - to know we are focusing on the right thing.)

+ Do you have any indications from IIS logs or APM monitoring of what may be happening?

+ The Polly policies provide the ability to track how they intervene in executions through the `onRetry`, `onBreak` etc delegates: if you want to track exactly how the Polly policies are intervening in production, hook logging into those `onPolicyDidSomething` delegates (if not already).  

+ What kind of volumes/throughput is the WCF service dealing with? (out of interest)

Finally, not quite clear on the topology: where are the Polly policies placed in relation to the caller, callee, load-balancer, etc?  (In microservices environment where I work, we some Polly both on the caller and callee sides, for different purposes.)

Hope some of this helps, and that we can help further with this!

**Comment by reisenberger** [comment]
@mieliespoor Also, if you want to share code sample of how the Polly policies are being used, of course we can look over that and comment if anything looks unusual.  Thanks!

**Comment by Ulriksen** [comment]
Are you sure that load balanced is the difference, could it be traffic/load related instead? 

When something hangs my first suspect is sync code calling async functions with with .wait or .result.

**Comment by reisenberger** [comment]
@mieliespoor Which version of Polly are you using?  And which kind of Policy?  I have identified a possible issue with the synchronous form of `WaitAndRetry(...)` for Polly v5.0.1 and v5.0.2 (only), and have a fix ready for Polly v5.0.3.  However, if you are using Polly v4.3.0 or earlier, this would not be relevant.

Thanks

**Comment by mieliespoor** [comment]
Hello,

It might be a load issue, because the load in production is significantly more than the test environments, I will however test this over the next few days.

I have implemented the Circuit breaker policy using v4.3.0 it seems.

Before I start with the testing, I will use the config flags to enable and disable the circuit breaker and see what effect it has got. Any comments on the code below would be appreciated.

To make it work in the WCF service, I had to implement dependency injection. IN our staging environment, it seems to work without any issue. The registration happens as follow (only way I could get it to work...):
`container
               .RegisterType<IConnxService, ConnxService>()
               .RegisterInstance<ICircuitBreakerSettings>(new CircuitBreakerSettings())
               .RegisterInstance<ICircuitBreaker>(new PollyCircuitBreaker(container.Resolve<ICircuitBreakerSettings>()))
               .RegisterInstance<IConnxDataHelper>(new ConnxDataHelper(container.Resolve<ICircuitBreaker>()));`

`public PollyCircuitBreaker(ICircuitBreakerSettings settings)
        {
            try
            {
                _policy = Policy
                         .Handle<Exception>()
                         .CircuitBreaker(settings.BreakOnNumberOfExceptions,
                             TimeSpan.FromSeconds(settings.BreakCircuitForSeconds),
                             OnCircuitOpen,
                             OnCircuitClosed);
            }
            catch (Exception ex)
            {
                _telemetryClient.TrackException(ex);
            }
        }

        public T Execute<T>(Func<T> action)
        {
            if (_policy == null)
            {
                return action.Invoke();
            }
            else
            {
                return _policy.Execute(action);
            }
        }`

As for samples:
`return _circuitBreaker.Execute(() =>
            {
                var success = true;
                var startTime = DateTime.UtcNow;
                var timer = System.Diagnostics.Stopwatch.StartNew();
                try
                {
                    ValidateConnxConnection();
                    using (CNXDataAdapter sqlda = new CNXDataAdapter(sql, connection))
                    using (DataSet ds = new DataSet())
                    {
                        ds.Locale = CultureInfo.InvariantCulture;
                        sqlda.Fill(ds);
                        if (ds != null && ds.Tables != null)
                        {
                            return ds.Tables[0];
                        }
                    }

                }
                catch (Exception)
                {
                    success = false;
                    throw;
                }
                finally
                {
                    timer.Stop();
                    TrackDependency("Connx", connection.Database, "CNXQuery", startTime, timer.Elapsed, success);
                }

                return null;
            });`

**Comment by reisenberger** [comment]
@mieliespoor Nothing in the code samples looks particularly unusual / likely to cause trouble.

(With Polly, as with any exception-handling, we sometimes advise handle only the exceptions you specifically _want_ to/can handle, rather than blanket every exception, to prevent policies handling exceptions not originally envisaged.  For instance, for I/O, we might advise handling only the expected I/O exceptions, rather than blanket `.Handle<Exception>`.  _Probably_ unrelated to the load-balancer/load issue you've raised (and with relevant logging, you can obv see what the exceptions are).  A blanket `.Handle<Exception>()` also is less of an issue with circuitbreaker, where handling unenvisaged exceptions is at worst going to trigger breaking the circuit more eagerly.  More of an issue with retry, where retrying on exceptions where no success could be expected, can be pointless and only consume resource.)

Interested to hear results of further investigations.

**Comment by reisenberger** [comment]
Also worth making sure - if the system is under high/near-capacity load - that the state-change delegates `OnCircuitOpen`, `OnCircuitClosed` aren't overly blocking, given their interaction with the locks held by the circuit-breaker described [here](https://github.com/App-vNext/Polly/wiki/Circuit-Breaker#delegates-on-transitions-of-circuit-state--polly-v40).  

(Somewhat 'clutching at straws' here though - you are likely to have stronger leads as to any possible issue, from your real diagnostics in your environment.)

**Comment by mieliespoor** [comment]
Hello,
Sorry I forgot to come back to this. I eventually did a big number of tests and found that the issues was actually not related to the circuit breaker code, but something else. It was actually a really old bug that surfaced when this code went to production. Took a bit to get it to reproduce, but once I were able to reproduce it, I could get it resolved.

This did however provide the opportunity to properly show how Polly works.

With regards to not catching a general 'Exception' - yeah, I'm against doing that. This was just to get it going and working. We do actually have a specific exception this needs to check.

In the `OnCircuitOpen` and `OnCircuitClosed ` methods, all that happens for the moment, is event logging to Application Insights - that won't hold anything up...
